api/pkg/controller/addoninstaller/controller.go

```
@@ -0,0 +1,248 @@
+				if !ok {
+					utilruntime.HandleError(fmt.Errorf("couldn't get obj from tombstone %#v", obj))
+					return
+				}
+				cluster, ok = tombstone.Obj.(*kubermaticv1.Cluster)
+				if !ok {
+					utilruntime.HandleError(fmt.Errorf("tombstone contained object that is not a Cluster %#v", obj))
+					return
+				}
+			}
+
+			c.enqueueCluster(cluster)
+		},
+	})
+
+	c.clusterLister = clusterInformer.Lister()
+	c.clusterSynced = clusterInformer.Informer().HasSynced
+
+	c.addonLister = addonInformer.Lister()
+	c.addonSynced = addonInformer.Informer().HasSynced
+
+	return c, nil
+}
+
+// If an clusterInformer triggers queuing, put the cluster into the workqeue
+func (c *Controller) enqueueCluster(cluster *kubermaticv1.Cluster) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(cluster)
+
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", cluster, err))
+		return
+	}
+	c.queue.Add(key)
+}
+
+// make API call to create an addon in the cluster
+func (c *Controller) createDefaultAddon(addon string, cluster *kubermaticv1.Cluster) {
+	gv := kubermaticv1.SchemeGroupVersion
+	ownerRef := *metav1.NewControllerRef(cluster, gv.WithKind("Cluster"))
+	glog.V(8).Infof("Create addon %s for the cluster %s\n", addon, cluster.Name)
+
+	if _, err := c.client.KubermaticV1().Addons(cluster.Status.NamespaceName).Create(&kubermaticv1.Addon{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            addon,
+			Namespace:       cluster.Status.NamespaceName,
+			OwnerReferences: []metav1.OwnerReference{ownerRef},
+		},
+		Spec: kubermaticv1.AddonSpec{
+			Name: addon,
+			Cluster: corev1.ObjectReference{
+				Name:       cluster.Name,
+				Namespace:  "",
+				UID:        cluster.UID,
+				APIVersion: cluster.APIVersion,
+				Kind:       "Cluster",
+			},
+		},
+	}); err != nil {
+		utilruntime.HandleError(fmt.Errorf("failed to create initial adddon %s for cluster %s: %v", addon, cluster.Name, err))
```

This function gets called from the sync handler, so we have a cluster context and also have a error handler in place: https://github.com/kubermatic/kubermatic/pull/1431/files#diff-b6caaf6a977e60d21f84c70f79230f29R189

Why not simply return the error?
I think we can directly do a `return err` and enrich the error on https://github.com/kubermatic/kubermatic/pull/1431/files#diff-b6caaf6a977e60d21f84c70f79230f29R240 with the addon name